File: 2021/15/main.hs
import qualified Data.Map.Strict as M
import qualified Data.Set        as S
import qualified Data.Array      as A
import Debug.Trace

type Grid  = A.Array Pos Risk
type Flood = M.Map Pos Risk
type Scan  = S.Set Pos
type Pos   = (Int, Int)
type Risk  = Int

parseInput :: String -> Grid
--parseInput :: String -> [(Pos, Risk)]
parseInput t = A.array ((0, 0), s t) $ g t
  where s t = ((length . lines) t - 1, (length . head . lines) t - 1)
        g   = foldMap (\(x,l) -> map (\(y,v) -> ((y, x), read [v])) l)
            . zip [0..] . map (zip [0..])
            . lines

bfs :: Grid -> Risk
bfs grid = loop (M.singleton (0, 0) 0) (S.singleton (0, 0))
  where
    end@(ex, ey) = snd $ A.bounds grid
    loop :: Flood -> Scan -> Risk
    loop flood scan
      | scan' == S.empty = flood M.! end
      | otherwise        = loop flood' $ scan'
      where
        (scan', flood') = S.foldr update (S.empty, flood) scan
        update :: Pos -> (Scan, Flood) -> (Scan, Flood)
        update pos (scan, flood) = (scan', flood')
          where
            flood' = foldr (uncurry M.insert) flood pr
            scan' = foldr S.insert scan $ map fst pr
            nb = neighbours pos
            rs = map ((+ val) . (grid A.!)) nb
            pr = filter (\(k, v) -> get k > v) $ zip nb rs
            val = get pos
            get :: Pos -> Risk
            get = maybe maxBound id . (flip M.lookup) flood
        set k = M.insert k
        neighbours (x, y) = filter inRange
                          $ [(x + 1, y), (x, y + 1), (x - 1, y), (x, y - 1)]
        inRange (x, y) = 0 <= x && x <= ex && 0 <= y && y <= ey

enlarge :: Int -> Grid -> Grid
enlarge n grid = A.array ((0, 0), (sx * n - 1, sy * n - 1)) a
  where
    a = [ ((mx * sx + x, my * sy + y), f mx my x y)
        | x <- [0..ex]
        , y <- [0..ey]
        , mx <- [0..n - 1]
        , my <- [0..n - 1]
        ]
    f mx my x y = (mx + my + grid A.! (x, y) - 1) `mod` 9 + 1
    (ex, ey) = snd $ A.bounds grid
    (sx, sy) = (ex + 1, ey + 1)

main = parseInput <$> readFile "input.txt"
   >>= mapM_ print . sequence [bfs, bfs . enlarge 5]

File: 2021/15/main.rs
use std::collections::HashSet;
use std::fs;
use std::mem;

struct Grid {
    width: usize,
    height: usize,
    data: Box<[usize]>,
}

impl Grid {
    fn parse_input(text: Vec<u8>) -> Self {
        let height = text.split(|&c| c == b'\n').count() - 1; // off by one, idk
        let width = text.split(|&c| c == b'\n').next().unwrap().len();
        let data = text
            .into_iter()
            .filter(|&c| c != b'\n')
            .map(|c| usize::from(c - b'0'))
            .collect();
        Self {
            height,
            width,
            data,
        }
    }

    fn bfs(&self) -> usize {
        let mut flood = Self {
            width: self.width,
            height: self.height,
            data: self.data.iter().map(|_| usize::MAX).collect(),
        };
		flood.data[0] = 0;
        let mut scan = HashSet::new();
        let mut scan_next = HashSet::new();
        scan.insert((0, 0));
        loop {
            if scan.is_empty() {
                return flood.get(self.width - 1, self.height - 1).unwrap();
            }
            for (x, y) in scan.drain() {
                let v = flood.get(x, y).unwrap();
                let mut f = |x, y| {
                    let v = v + self.get(x, y).unwrap();
                    if v < flood.get(x, y).unwrap() {
                        flood.set(x, y, v).unwrap();
						scan_next.insert((x, y));
                    }
                };
                (x < self.width - 1).then(|| f(x + 1, y));
                (y < self.height - 1).then(|| f(x, y + 1));
                (x > 0).then(|| f(x - 1, y));
                (y > 0).then(|| f(x, y - 1));
            }
            mem::swap(&mut scan, &mut scan_next);
        }
    }

    fn enlarge(self, n: usize) -> Self {
        let (sx, sy) = (self.width, self.height);
        let f = |x, y, mx, my| -> usize {
			(mx + my + self.get(x, y).unwrap() - 1) % 9 + 1
		};
        Self {
            width: self.width * n,
            height: self.height * n,
            data: (0..n).flat_map(move |my| {
					(0..sy).flat_map(move |y| {
						(0..n).flat_map(move |mx| {
							(0..sx).map(move |x| {
								f(x, y, mx, my)
							})
						})
                    })
                })
                .collect(),
        }
    }

    fn get(&self, x: usize, y: usize) -> Option<usize> {
        (x < self.width && y < self.height).then(|| self.data[y * self.width + x])
    }

    fn set(&mut self, x: usize, y: usize, value: usize) -> Option<()> {
        (x < self.width && y < self.height).then(|| self.data[y * self.width + x] = value)
    }
}

fn main() {
    let grid = Grid::parse_input(fs::read("input.txt").unwrap());
    println!("{}", grid.bfs());
    println!("{}", grid.enlarge(5).bfs());
}
